src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/VideoLinkBookingMigrationService.kt

```diff
@@ -0,0 +1,89 @@
+import uk.gov.justice.digital.hmpps.whereabouts.services.VideoLinkBookingMigrationService.Outcome.MIGRATED_NO_MODIFICATIONS
+import uk.gov.justice.digital.hmpps.whereabouts.services.VideoLinkBookingMigrationService.Outcome.POST_DELETED
+import uk.gov.justice.digital.hmpps.whereabouts.services.VideoLinkBookingMigrationService.Outcome.PRE_DELETED
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.LocalTime
+
+@Service
+@Transactional
+class VideoLinkBookingMigrationService(
+
+  val videoLinkBookingRepository: VideoLinkBookingRepository,
+  val videoLinkAppointmentRepository: VideoLinkAppointmentRepository,
+  val prisonApiService: PrisonApiService,
+) {
+  enum class Outcome { MIGRATED_NO_MODIFICATIONS, BOOKING_DELETED, PRE_DELETED, POST_DELETED }
+
+  fun migrateFromNomis(fromDate: LocalDate, toDate: LocalDate): VideoLinkAppointmentMigrationResponse {
+    var videoLinkAppointments = videoLinkAppointmentRepository.findAllByStartDateTimeBetweenAndHearingTypeIs(
+      LocalDateTime.of(fromDate, LocalTime.MIN),
+      LocalDateTime.of(toDate, LocalTime.MAX),
+      HearingType.MAIN
+    )
+    val outcomes = videoLinkAppointments.flatMap { v -> updateVideoLink(v.videoLinkBooking) }
+
+    return VideoLinkAppointmentMigrationResponse(
+      videoLinkAppointments.size,
+      outcomes.groupingBy { it }.eachCount()
+    )
+  }
+
+  fun updateVideoLink(videoLinkBooking: VideoLinkBooking): List<Outcome> {
+    val mainAppointment = videoLinkBooking.appointments[HearingType.MAIN]
+    val preAppointment = videoLinkBooking.appointments[HearingType.PRE]
+    val postAppointment = videoLinkBooking.appointments[HearingType.POST]
+
+    val outcomes = mutableListOf<Outcome>()
+    val nomisMainAppointment = updateAppointment(mainAppointment!!)
+    if (nomisMainAppointment == null) {
+      log.info("Deleting booking: ${videoLinkBooking.id}")
+      videoLinkBookingRepository.delete(videoLinkBooking)
+      outcomes.add(BOOKING_DELETED)
+    } else {
+      videoLinkBooking.prisonId = nomisMainAppointment.agencyId
+      videoLinkBooking.comment = nomisMainAppointment.comment
+
+      if (preAppointment?.let { updateAppointment(it) } == null) {
+        videoLinkBooking.appointments.remove(HearingType.PRE)
+        outcomes.add(PRE_DELETED)
+      }
+      if (postAppointment?.let { updateAppointment(it) } == null) {
+        videoLinkBooking.appointments.remove(HearingType.POST)
+        outcomes.add(POST_DELETED)
+      }
+      videoLinkBookingRepository.save(videoLinkBooking)
+    }
+    return if (outcomes.isEmpty()) listOf(MIGRATED_NO_MODIFICATIONS) else outcomes
+  }
+
+  private fun updateAppointment(appointment: VideoLinkAppointment): PrisonAppointment? {
+    val nomisAppointment = prisonApiService.getPrisonAppointment(appointment.appointmentId)
+    if (nomisAppointment == null) {
+      log.info("Deleting appointment: ${appointment.appointmentId}")
+      return null
```

Comment:
think we had to rewrite this otherwise it reports outcomes when there is no preappointment on the booking.
https://github.com/ministryofjustice/whereabouts-api/pull/513/files